fastcrypto-zkp/src/bn254/zk_login.rs

```diff
@@ -123,380 +170,227 @@ impl fmt::Display for SupportedKeyClaim {
-pub struct JWTHeader {
-    alg: String,
-    kid: String,
-    typ: String,
-}
-
-impl ParsedMaskedContent {
-    /// Parse the masked content bytes into a [struct ParsedMaskedContent].
-    /// payload_start_index, payload_len, num_sha2_blocks are used for
-    /// validation and parsing.
-    pub fn new(
-        masked_content: &[u8],
-        payload_start_index: u16,
-        payload_len: u16,
-        num_sha2_blocks: u8,
-    ) -> Result<Self, FastCryptoError> {
-        // Verify the bytes after 64 * num_sha2_blocks should be all 0s.
-        if !masked_content
-            .get(64 * num_sha2_blocks as usize..)
-            .ok_or(FastCryptoError::InvalidInput)?
-            .iter()
-            .all(|&x| x == 0)
-        {
-            return Err(FastCryptoError::GeneralError(
-                "Incorrect payload padding".to_string(),
-            ));
-        }
-
-        let masked_content_tmp = masked_content
-            .get(..64 * num_sha2_blocks as usize)
-            .ok_or(FastCryptoError::InvalidInput)?;
+    /// Get address seed string.
+    pub fn get_address_params(&self) -> AddressParams {
+        AddressParams::new(self.get_iss().to_owned(), self.get_aud().to_owned())
+    }
 
-        // Verify the byte at payload start index is indeed b'.'.
-        if payload_start_index < 1
-            || masked_content_tmp.get(payload_start_index as usize - 1) != Some(&b'.')
-        {
-            return Err(FastCryptoError::GeneralError(
-                "Incorrect payload index for separator".to_string(),
-            ));
+    /// Calculate the poseidon hash from selected fields from inputs, along with the ephemeral pubkey.
+    pub fn calculate_all_inputs_hash(
+        &self,
+        eph_pubkey_bytes: &[u8],
+        modulus: &[u8],
+        max_epoch: u64,
+    ) -> Result<Vec<Bn254Fr>, FastCryptoError> {
+        if self.header_base64.len() > MAX_HEADER_LEN as usize {
+            return Err(FastCryptoError::GeneralError("Header too long".to_string()));
         }
 
-        let header = parse_and_validate_header(
-            masked_content_tmp
-                .get(0..payload_start_index as usize - 1)
-                .ok_or_else(|| {
-                    FastCryptoError::GeneralError(
-                        "Invalid payload index to parse header".to_string(),
-                    )
-                })?,
-        )?;
+        let mut poseidon = PoseidonWrapper::new();
+        let addr_seed = to_field(&self.address_seed)?;
 
-        // Parse the jwt length from the last 8 bytes of the masked content.
-        let jwt_length_bytes = masked_content_tmp
-            .get(masked_content_tmp.len() - 8..)
-            .ok_or_else(|| FastCryptoError::GeneralError("Invalid last 8 bytes".to_string()))?;
-        let jwt_length = calculate_value_from_bytearray(jwt_length_bytes);
+        let (first_half, second_half) = eph_pubkey_bytes.split_at(eph_pubkey_bytes.len() / 2);
```

Comment:
There is a mismatch between the logic in TS code and this. In the TS code, the splitting is deterministic, i.e., we compute `eph_pubkey_bytes / 2^128` and `eph_pubkey_bytes % 2^128`. If the public key is bigger than 256 bits, then I think we have a bug at hand. 

Could you update this to always split at 128 bits, i.e., 16 bytes?

Reply:
revised - it should be 17 since the index here is exclusive!